src/main/java/javaslang/collection/List.java

```diff
@@ -426,7 +426,15 @@ default int lastIndexOf(T element) {
     @Override
     default Tuple2<List<T>, List<T>> partition(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        return Tuple.of(filter(predicate), filter(predicate.negate()));
+        List<T> left = nil(), right = nil();
+        for (T t : this) {
+            if (predicate.test(t)) {
+                left = left.prepend(t);
+            } else {
+                right = right.prepend(t);
+            }
+        }
+        return Tuple.of(left.reverse(), right.reverse());
```

Comment:
Yes, you're right :-) It is better than the solution before, which used 2 x filter(), which uses foldLeft(), which also calls reverse(), i.e. 4 passes in sum. There is no improvement when using ArrayList for building lists because List.ofAll(Iterable) has to be called to convert the ArrayList to a javaslang.collection.List, which internally also needs to call reverse(). I think there is no better solution for a single-linked list.